src/compiler/checker.ts

```diff
@@ -26982,18 +26982,38 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 const inferredCovariantType = inference.candidates ? getCovariantInference(inference, context.signature) : undefined;
                 const inferredContravariantType = inference.contraCandidates ? getContravariantInference(inference) : undefined;
                 if (inferredCovariantType || inferredContravariantType) {
-                    // If we have both co- and contra-variant inferences, we prefer the co-variant inference if it is not 'never',
-                    // all co-variant inferences are assignable to it (i.e. it isn't one of a conflicting set of candidates), it is
-                    // assignable to some contra-variant inference, and no other type parameter is constrained to this type parameter
-                    // and has inferences that would conflict. Otherwise, we prefer the contra-variant inference.
-                    // Similarly ignore co-variant `any` inference when both are available as almost everything is assignable to it
-                    // and it would spoil the overall inference.
+                    // ideally all inferences would be tried out in a ranked order but that's too costly

```

Comment:
I rewritten this comment based on the excellent comment by @weswigham [here](https://github.com/microsoft/TypeScript/pull/57909#pullrequestreview-2119045380). I tried to be concise though as this is already pretty long and Wes' comment was even longer 😉 Feel free to suggest any improvement to this to make it better. I write better in TS than in English.